Separate conversion pass for `torch -> tcp.custom_op` to support both kernel and codegen approaches

sjain-stanford opened the pull request:
This allows switching between codegen vs kernels approaches, for torch ops that can lower to either tcp ops (codegen) or tcp.custom_op (kernel). Addresses https://github.com/cruise-automation/mlir-tcp/pull/17#discussion_r1382303711 using [Option 2](https://github.com/cruise-automation/mlir-tcp/pull/17#discussion_r1382436616).

sanjoy commented on include/mlir-tcp/Conversion/TorchToTcp/TorchToTcpCustomOp.h:
```diff
@@ -0,0 +1,29 @@
+
+std::unique_ptr<OperationPass<func::FuncOp>>
+createConvertTorchToTcpCustomOpPass();
+std::unique_ptr<OperationPass<func::FuncOp>>
```
Nit: add newline between function declarations.

sanjoy commented on lib/Conversion/TorchToTcp/TorchToTcp.cpp:
```diff
@@ -97,6 +94,11 @@ class ConvertTorchToTcp : public ConvertTorchToTcpBase<ConvertTorchToTcp> {
 
+std::unique_ptr<OperationPass<func::FuncOp>> createConvertTorchToTcpPass() {
+  llvm::ArrayRef<std::string> emptyArrayRef;
+  return std::make_unique<ConvertTorchToTcp>(/*convertTorchOps=*/emptyArrayRef);
```
Can this be `std::make_unique<ConvertTorchToTcp>({})`

sjain-stanford replied:
I initially had that, but hit this compiler error:
```
lib/Conversion/TorchToTcp/TorchToTcp.cpp:99:10: error: no matching function for call to 'make_unique'
  return std::make_unique<ConvertTorchToTcp>({});
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ 
/usr/lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/unique_ptr.h:961:5: note: candidate template ignored: substitution failure [with _Tp = mlir::tcp::(anonymous namespace)::ConvertTorchToTcp]: deduced incomplete pack <(no value)> for template parameter '_Args'
    make_unique(_Args&&... __args)                                                                                                                                                                     
    ^                                           
    ```

sjain-stanford commented:
@navahgar I am about to land this under the assumption that it addresses your comment about tcp.gather codegen. Please let me know if it doesn't or if you have any other feedback, and I'll be glad to take it up in a follow-on.
